Issue #18017 Operators * / for Size and SizeF

WinCPP opened the pull request:
Fixes #18017 .

@onovotny @karelz @terrajobst @mellinoe Kindly review.

WinCPP commented:
@dotnet-bot please run Innerloop CentOS7.1 Debug x64 Build and Test

WinCPP commented:
@dotnet-build-bot Please run Innerloop CentOS7.1 Debug x64 Build and Test

WinCPP commented:
@dotnet-bot Innerloop CentOS7.1 Debug x64 Build and Test

onovotny commented on src/System.Drawing.Primitives/src/System/Drawing/Size.cs:
```diff
@@ -190,5 +240,25 @@ public int Height
         ///    </para>
         /// </summary>
         public override string ToString() => "{Width=" + _width.ToString() + ", Height=" + _height.ToString() + "}";
```
Minor nit - can use string interpolation instead of adding strings with + and ToString()

WinCPP replied:
Yup got it... but, hmmm..., do I make this change in this PR since this is existing code...? Because I learnt over last few months that the scope of the changes in PR should be restricted to the issue... 🤔

onovotny replied:
Didn't realize it was existing, looked green in the diff. That's a better question for @mellinoe then.

mellinoe replied:
Just keep it as-is if it is not part of your changes.

onovotny commented on src/System.Drawing.Primitives/src/System/Drawing/SizeF.cs:
```diff
@@ -184,6 +209,16 @@ public float Height
         ///    </para>
         /// </summary>
         public override string ToString() => "{Width=" + _width.ToString() + ", Height=" + _height.ToString() + "}";
```
Minor nit - can use string interpolation instead of adding strings with + and ToString()

WinCPP replied:
same as above...

WinCPP commented:
@karelz netfx build and test is failing. I think I need to make changes to somewhere, but I forget where... Kindly advise / redirect...

karelz commented:
Here's the error from the link above: `error CS0019: Operator '*' cannot be applied to operands of type 'SizeF' and 'float'`
Here's how you can build locally: https://github.com/dotnet/corefx/blob/c2588415e91ca1d44885ac0bbe9bf8268adc4b48/Documentation/project-docs/developer-guide.md (`build -framework=netfx`)

If that doesn't help, I can loop in experts (Eric M. is already on the PR).

tannergooding commented on src/System.Drawing.Primitives/src/System/Drawing/Size.cs:
```diff
@@ -74,6 +74,56 @@ public Size(int width, int height)
+        /// <param name="size">Dividend of type <see cref="Size"/>.</param>
+        /// <param name="divisor">Divisor of type <see cref="int"/>.</param>
+        /// <returns>Result of type <see cref="Size"/>.</returns>
+        public static Size operator /(Size size, int divisor) =>
```
nit: We should make this do multiplication by reciprocal, as we do in the System.Numerics.Vector* types.

WinCPP replied:
I'll update the same and hence the test cases too... I believe without that the equality check would fail...?

tannergooding commented on src/System.Drawing.Primitives/src/System/Drawing/Size.cs:
```diff
@@ -74,6 +74,56 @@ public Size(int width, int height)
+        /// <param name="size">Dividend of type <see cref="Size"/>.</param>
+        /// <param name="divisor">Divisor of type <see cref="int"/>.</param>
+        /// <returns>Result of type <see cref="SizeF"/>.</returns>
+        public static SizeF operator /(Size size, float divisor) =>
```
nit: Same as above, we should do multiplication by reciprocal

tannergooding commented on src/System.Drawing.Primitives/src/System/Drawing/SizeF.cs:
```diff
@@ -84,6 +84,31 @@ public SizeF(float width, float height)
+        /// <param name="size">Dividend of type <see cref="SizeF"/>.</param>
+        /// <param name="divisor">Divisor of type <see cref="int"/>.</param>
+        /// <returns>Result of type <see cref="SizeF"/>.</returns>
+        public static SizeF operator /(SizeF size, float divisor) =>
```
nit: Same as above, we should do multiplication by reciprocal

WinCPP commented:
Yup I'm getting that error on my local build as well, using the -framework=netfx; same as I saw on the build machines. Looks like I have to expose the new operators in some .cs file for netfx to work...?

tannergooding commented on src/System.Drawing.Primitives/tests/SizeFTests.cs:
```diff
@@ -77,6 +77,57 @@ public void ArithmeticTest(float width, float height)
         }
 
         [Theory]
+        [InlineData(1000.234, 2400.933)]
```
Shouldn't all these values be float, rather than double (i.e. `1000.234f` instead of `1000.234`)?

tannergooding commented:
@WinCPP, not positive here, but I believe you need to mark the new APIs as netcore only. They won't be useable on Desktop until some point in the future when the changes are merged back into System.Drawing.dll.

karelz commented:
> Looks like I have to expose the new operators in some .cs file for netfx to work...?

Oh, I think that you need to make the tests only for netcoreapp20, that's IMO what is happening. It is not super-well documented, but you can check some past PRs which did that. Some hints might be [here](https://github.com/dotnet/corefx/wiki/Contributions#expose-api-across-coreclrcorefx).

@mellinoe @weshaggard @ericstj can you please help provide guidance? Let's capture it [in the docs](https://github.com/dotnet/corefx/wiki/Contributions#expose-api-across-coreclrcorefx) for future. Thanks!

ericstj commented:
Yes, cross-compile the tests just as you would any other project in corefx that needs to be framework specific.  Docs are here: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/project-guidelines.md#aggregate-build-properties

mellinoe commented on src/System.Drawing.Primitives/tests/SizeTests.cs:
```diff
@@ -109,6 +109,125 @@ public void ArithmeticTest(int width, int height)
+        {
+            if (value2 == 2400.993)
+            {
+                Console.WriteLine("This is it {0}", value1);
```
Debug code?

WinCPP replied:
Oops.

mellinoe commented on src/System.Drawing.Primitives/tests/SizeTests.cs:
```diff
@@ -109,6 +109,125 @@ public void ArithmeticTest(int width, int height)
+            Size size = new Size(100, 100);
+            Assert.Throws<DivideByZeroException>(() => size / 0);
+
+            SizeF expectedSizeF = new SizeF(100 / (float)0.0, 100 / (float)0.0);
```
Can you not just use float literals?

mellinoe replied:
Alternatively, can this just be `float.PositiveInfinity` ?

mellinoe commented:
I added a few more comments, and @tannergooding's suggestions should be followed as well.

@alexperovich Do you want to take a look as well?

WinCPP commented:
@mellinoe Between, I just noticed in #7175 that `left` and `right` are the names as per design guidelines and multiplicand, multiplier, dividend, divisor are not (discussion [here](https://github.com/dotnet/corefx/issues/7175#issuecomment-262366418)). I guess I should change to left / right? I had removed that because when it came to intellisense I thought it might sound a bit odd...

tannergooding commented:
@WinCPP, I believe that `left` and `right` are the names, but you can clarify the description (indicating what is the `dividend` and what is the `divisor`) in the doc comments (which IntelliSense will display in your IDE).

mellinoe commented:
Yes, go with `left` and `right`.

WinCPP commented:
@ericstj Just to make sure my understanding is correct ...

The link tells how to restrict a project to specific framework. In this case, the set of new test cases are to be restricted to netcoreapp specific. So I have two options - move them to different project specific to netcoreapp or I think, I saw somewhere, move them to a .cs file that come in only on netcoreapp framework...? Appreciate your inputs.

ericstj commented:
Don't use a seperate project.  Add an entry to the configurations.props for the tests for netcoreapp.  This will cause the test to cross-compile for netcoreapp.  Then you have a choice to seperate the tests into a seperate source file and use a condition in the project file, or use ifdefs in the source file to condition them out.  There are several examples of this throughout the repo.  Just look at any test that has more than one configuration in its configurations.props to get an idea of how folks use it.

tannergooding commented:
@WinCPP, the `System.MathF` tests are one example: https://github.com/dotnet/corefx/tree/master/src/System.Runtime.Extensions/tests/System

The file is called `MathF.netcoreapp.cs` and is included in a conditioned ItemGroup:
```XML
<ItemGroup Condition="'$(TargetGroup)'=='netcoreapp'">
    <Compile Include="System\MathF.netcoreapp.cs" />
</ItemGroup>
```

karelz commented:
@tannergooding could you please help us craft the docs for new contributors pointing out the example and all its necessary pieces? https://github.com/dotnet/corefx/wiki/Contributions#expose-api-across-coreclrcorefx

tannergooding commented:
@karelz, sure thing. I should have some time later this evening to do so.

karelz commented:
Perfect! Thank you! ... BTW: I asked for help with these docs also on [corefx-contrib-docs gitter](https://gitter.im/dotnet/corefx-contrib-docs?at=5925ba81eec422e415e564c2) - you might want to check there first if someone didn't grab it ...

tannergooding commented on src/System.Drawing.Primitives/src/System/Drawing/Size.cs:
```diff
@@ -74,6 +74,62 @@ public Size(int width, int height)
+        public static Size operator /(Size left, int right)
+        {
+            int invDiv = 1 / right;
+            return new Size(unchecked(left._width * invDiv), unchecked(left._height * invDiv));
```
why not just call `Multiply(left, invDiv)`?

tannergooding replied:
Actually, this one we shouldn't do multiplication by reciprocal (my fault for tagging this in the original pass). Integers can't do this optimization because 1 / x where |x| > 0 always results in 0. Multiplication by reciprocal is only a valid optimization when dealing with floating-point numbers.

mellinoe replied:
We do some similar manual inlining in other types. It's been found to be faster in some cases. This is a legacy type, so I don't really think high-performance is a priority, but I'm fine with the way it is.

tannergooding commented on src/System.Drawing.Primitives/src/System/Drawing/Size.cs:
```diff
@@ -74,6 +74,62 @@ public Size(int width, int height)
+        public static SizeF operator /(Size left, float right)
+        {
+            float invDiv = 1.0f / right;
+            return new SizeF(left._width / invDiv, left._height / invDiv);
```
These should be * not /. Also, just calling `Multiply(left, invDiv)` might be clearer

mellinoe replied:
+1. Please add a test that covers this case.

WinCPP replied:
The test case with input [Int.Max/Min, Float.Max/Min] (4 combinations) failed as soon as I corrected / to *. But I'm surprised why all of them passed when it was /....

tannergooding commented on src/System.Drawing.Primitives/src/System/Drawing/SizeF.cs:
```diff
@@ -84,6 +84,34 @@ public SizeF(float width, float height)
+        public static SizeF operator /(SizeF left, float right)
+        {
+            float invDiv = 1.0f / right;
+            return new SizeF(unchecked(left._width * invDiv), unchecked(left._height * invDiv));
```
`Multiply(left, invDiv)` might be clearer

tannergooding commented on src/System.Drawing.Primitives/tests/SizeTests.netcoreapp.cs:
```diff
@@ -0,0 +1,132 @@
+
+            float invDiv = 1.0f / value2;
+
+            expected = new SizeF(value1 * invDiv, value1 * invDiv);
```
The tests should do normal division for the expected result, not multiplication by reciprocal. This ensures that any errors in the operator (due to possible rounding errors, etc) are exposed.

WinCPP replied:
I think I tried asking that in [this](https://github.com/dotnet/corefx/pull/20052#discussion_r117809814) comment.

Between, the test case fails if I do not use reciprocal multiplication in the test case. I could not understand why we would want to use an approach outside (test case) which is different than inside the operator. Because the operator is just a syntactic sugar and supposed to behave on _whole_ in identical way as it would behave with _parts_; and really doesn't change the mathematics...?

For the following inputs.
```
        [InlineData(int.MaxValue, float.MaxValue)]
        [InlineData(int.MaxValue, float.MinValue)]
        [InlineData(int.MinValue, float.MaxValue)]
        [InlineData(int.MinValue, float.MinValue)]
```
The output is this,
```  D:\WinCPP\CoreFx\bin/AnyOS.AnyCPU.Debug/System.Drawing.Primitives.Tests/netcoreapp//RunTests.cmd D:\WinCPP\CoreFx\bin/testhost/netcoreapp-Windows_NT-Debug-x64/
  Using D:\WinCPP\CoreFx\bin\testhost\netcoreapp-Windows_NT-Debug-x64\ as the test runtime folder.
  Executing in D:\WinCPP\corefx\bin\AnyOS.AnyCPU.Debug\System.Drawing.Primitives.Tests\netcoreapp\
  Running tests... Start time:  1:18:29.82
  xUnit.net console test runner (64-bit .NET Core)
  Copyright (C) 2014 Outercurve Foundation.

  Discovering: System.Drawing.Primitives.Tests
  Discovered:  System.Drawing.Primitives.Tests
  Starting:    System.Drawing.Primitives.Tests
     System.Drawing.PrimitivesTests.SizeTests.DivideTestSizeFloat(value1: 2147483647, value2: 3.402823E+38) [FAIL]
        Assert.Equal() Failure
        Expected: {Width=6.310888E-30, Height=6.310888E-30}
        Actual:   {Width=6.310887E-30, Height=6.310887E-30}
        Stack Trace:
           D:\WinCPP\CoreFx\src\System.Drawing.Primitives\tests\SizeTests.netcoreapp.cs(128,0): at System.Drawing.PrimitivesTests.SizeTests.DivideTestSizeFloat(Int32 value1, Singl
  e value2)
     System.Drawing.PrimitivesTests.SizeTests.DivideTestSizeFloat(value1: 2147483647, value2: -3.402823E+38) [FAIL]
        Assert.Equal() Failure
        Expected: {Width=-6.310888E-30, Height=-6.310888E-30}
        Actual:   {Width=-6.310887E-30, Height=-6.310887E-30}
        Stack Trace:
           D:\WinCPP\CoreFx\src\System.Drawing.Primitives\tests\SizeTests.netcoreapp.cs(128,0): at System.Drawing.PrimitivesTests.SizeTests.DivideTestSizeFloat(Int32 value1, Singl
  e value2)
     System.Drawing.PrimitivesTests.SizeTests.DivideTestSizeFloat(value1: -2147483648, value2: 3.402823E+38) [FAIL]
        Assert.Equal() Failure
        Expected: {Width=-6.310888E-30, Height=-6.310888E-30}
        Actual:   {Width=-6.310887E-30, Height=-6.310887E-30}
        Stack Trace:
           D:\WinCPP\CoreFx\src\System.Drawing.Primitives\tests\SizeTests.netcoreapp.cs(128,0): at System.Drawing.PrimitivesTests.SizeTests.DivideTestSizeFloat(Int32 value1, Singl
  e value2)
     System.Drawing.PrimitivesTests.SizeTests.DivideTestSizeFloat(value1: -2147483648, value2: -3.402823E+38) [FAIL]
        Assert.Equal() Failure
        Expected: {Width=6.310888E-30, Height=6.310888E-30}
        Actual:   {Width=6.310887E-30, Height=6.310887E-30}
        Stack Trace:
           D:\WinCPP\CoreFx\src\System.Drawing.Primitives\tests\SizeTests.netcoreapp.cs(128,0): at System.Drawing.PrimitivesTests.SizeTests.DivideTestSizeFloat(Int32 value1, Singl
  e value2)
  Finished:    System.Drawing.Primitives.Tests
```

tannergooding replied:
Floating-point values are themselves inexact (in most cases). As an example: `int.MaxValue` = `2147483647`. However, `(float)(int.MaxValue)` = `2147484000`.

So, while in real-math 1 / 3 and 1 * (1 / 3) are the same, they might not produce exactly the same value with IEEE floating-point numbers.

Generally, for single-precision values, 6-9 significant digits is considered acceptable (its 15-17 digits for double-precision values) for the final computed result (this is 6-9 significant digits of precision between each operation, so 3 * 2 / 1 is two separate operations. So the more operations performed, the more likely a number could diverge, especially when dealing with extremely large or extremely small numbers).

tannergooding replied:
xUnit actually has a special Assert.Equal() overload for double-precision to help handle these comparisons, but it doesn't have an equivalent for single-precision values yet.

WinCPP replied:
Bringing back the comment that I had mentioned above previously... In case of division for combinations (Size, float) and (SizeF, float), not doing reciprocal multiplication in test case is not giving results identical to operator /. E.g.

```
     System.Drawing.PrimitivesTests.SizeTests.DivideTestSizeFloat(value1: 2147483647, value2: 3.402823E+38) [FAIL]
        Assert.Equal() Failure
        Expected: {Width=6.310888E-30, Height=6.310888E-30}
        Actual:   {Width=6.310887E-30, Height=6.310887E-30}
        Stack Trace:
           D:\WinCPP\CoreFx\src\System.Drawing.Primitives\tests\SizeTests.netcoreapp.cs(125,0): at System.Drawing.PrimitivesTests.SizeTests.DivideTestSizeFloat(Int32 value1, Single value2)

     System.Drawing.PrimitivesTest.SizeFTests.DivideTestSizeFloat(value1: -3.402823E+38, value2: -3.402823E+38) [FAIL]
        Assert.Equal() Failure
        Expected: {Width=1, Height=1}
        Actual:   {Width=0.9999999, Height=0.9999999}
        Stack Trace:
           D:\WinCPP\CoreFx\src\System.Drawing.Primitives\tests\SizeFTests.netcoreapp.cs(64,0): at System.Drawing.PrimitivesTest.SizeFTests.DivideTestSizeFloat(Single value1, Single value2)
```

I thought that the `operator /` was just a syntactic sugar that worked on _whole_ in a way identical to how it would work on _parts_... But by using different approaches inside operator and in test case, are we ending up validating floating point division...? The failures are for these combinations,
```
operator / in Size
        [InlineData(int.MaxValue, float.MaxValue)]
        [InlineData(int.MaxValue, float.MinValue)]
        [InlineData(int.MinValue, float.MaxValue)]
        [InlineData(int.MinValue, float.MinValue)]

operator / in SizeF
        [InlineData(float.MaxValue, float.MaxValue)]
        [InlineData(float.MaxValue, float.MinValue)]
        [InlineData(float.MinValue, float.MaxValue)]
        [InlineData(float.MinValue, float.MinValue)]

tannergooding replied:
As mentioned, with extremely large or extremely large numbers, variations in how the math is performed will impact the actual result.

6-9 significant digits of accuracy is generally considered acceptable (for single-precision), so provided the result is accurate to within that precision (in the cases you listed above, it is), then the result is fine.

There is an Assert.Equal function here: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/tests/System/MathF.netcoreapp.cs#L31 that does the appropriate comparisons. It essentially boils down to the below (with a bunch of pre-checks to handle special values, such as NaN, -0, or +/- Infinity):
```C#
var delta = Math.Abs(actual - expected);

if (delta > variance)
{
    throw new EqualException(ToStringPadded(expected), ToStringPadded(actual));
}
```

WinCPP replied:
@tannergooding looks like our comments crossed each other. Please see my comment above.

tannergooding replied:
@mellinoe might be able to comment if we want to put the `Assert.Equal` function into some shared location or if we maybe want to handle this a different way...

WinCPP replied:
Yup I get the point about precision. However, my question was about operator / and operator * being more of convenience operators (syntactic sugars). Considering that they indeed are, wouldn't it be ideal to have identical behavior inside and outside the operator? Like crossing to opposite corner on a rectangle via top right or bottom left corner. Here those two being the operations - construction and division...?

tannergooding replied:
@WinCPP, the purpose of doing the two separate operations (multiply by reciprocal in the product code and actual division in the test code) is to validate that we didn't mess up the operator by performing some optimization or refactoring.

We want to multiply by reciprocal in the product code because it is faster.

We want to divide in the test to validate that our refactoring/optimization didn't break anything.

In this case, we end up with a slightly different result for some cases. However, the minor variations in the final result is generally considered acceptable (especially in code that deals with general purpose graphics). The general allowed variance is 6-9 significant digits, so provided our final result (even with the optimization) is within that range, we are still good.

WinCPP replied:
@tannergooding hmmm, I think I still tend to disagree. Because the focus of the test is to validate behavior or operator / and not that of floating point division. For example the ArithmeticTest in `Size`. The `Size.Add` and `operator +` is tested in same whole - part concept way and not in terms of other arithmetic operations such as `a + b` being tested as `a - (-1 * b)`...

Coming back to `operator /`, whatever mess might happen by performing optimization or refactoring would happen identically inside and outside and hence behavior of operator / would still be consistent. But, probably, there would be another test case, say, on floating point type that would validate result of division being within acceptable limits vis-a-vis reciprocal multiplication... which would flag off an error should there be any discrepancy due to any issues...

tannergooding replied:
I'll let @mellinoe have the final call here. I tend to think that, optimizations (even micro-optimizations) should be left out of tests entirely. And when testing floating-point values, you should always be testing that it is within an `allowedVariance`, due to the nature of IEEE floating-point values themselves.

mellinoe replied:
IMO, the simplest option is to just remove the reciprocal multiplication entirely. I consider this stuff all legacy (it only exists for compatibility), so we don't need to have micro-optimizations like this in it.

WinCPP replied:
@mellinoe sorry for late reply. Been stuck with some urgent work. I will update all the code and push it for review in next 2 - 3 days.

tannergooding commented on src/System.Drawing.Primitives/tests/SizeTests.netcoreapp.cs:
```diff
@@ -0,0 +1,132 @@
+            Size size = new Size(value1, value1);
+            Size expected;
+
+            int invDiv = 1 / value2;
```
The tests should do normal division for the expected result, not multiplication by reciprocal. This ensures that any errors in the operator (due to possible rounding errors, etc) are exposed.

WinCPP replied:
Deleted previous comment... it was due to bug that you pointed [here](https://github.com/dotnet/corefx/pull/20052#discussion_r118341516)

tannergooding commented on src/System.Drawing.Primitives/tests/SizeTests.netcoreapp.cs:
```diff
@@ -0,0 +1,132 @@
+            Size size = new Size(100, 100);
+            Assert.Throws<DivideByZeroException>(() => size / 0);
+
+            float invDiv = 1 / 0.0f;
```
The tests should do normal division for the expected result, not multiplication by reciprocal. This ensures that any errors in the operator (due to possible rounding errors, etc) are exposed.

tannergooding commented on src/System.Drawing.Primitives/tests/SizeFTests.netcoreapp.cs:
```diff
@@ -0,0 +1,69 @@
+        public void DivideTestSizeFloat(float value1, float value2)
+        {
+            SizeF size = new SizeF(value1, value1);
+            float invDiv = 1.0f / value2;
```
The tests should do normal division for the expected result, not multiplication by reciprocal. This ensures that any errors in the operator (due to possible rounding errors, etc) are exposed.

mellinoe commented on src/System.Drawing.Primitives/src/System/Drawing/Size.cs:
```diff
@@ -74,6 +74,62 @@ public Size(int width, int height)
         public static Size operator -(Size sz1, Size sz2) => Subtract(sz1, sz2);
 
         /// <summary>
+        /// Multiplication operator that multiplies <see cref="Size"/> by an <see cref="int"/> producing <see cref="Size"/>.
```
Can you remove "<foo> operator that" from all of these XML comments? It is a little redundant. This can just be `Multiplies a <see cref="Size"/> by an <see cref="int"/>, producing <see cref="Size"/>.`

mellinoe commented on src/System.Drawing.Primitives/src/System/Drawing/Size.cs:
```diff
@@ -190,5 +246,25 @@ public int Height
         public override string ToString() => "{Width=" + _width.ToString() + ", Height=" + _height.ToString() + "}";
+
+        /// <summary>
+        /// Helper method for operator * (multiply) that accepts type <see cref="Size"/> and <see cref="int"/>,
```
I would just have a description of what the method actually does; you don't need to describe that it is a "helper" for another operator. The description can be essentially identical to the operator's description.

mellinoe replied:
Same applies to the other named operator methods.

karelz commented:
@WinCPP what's the status of this PR? Do you plan to address the feedback? (if you don't have time right now, that's ok, just let us know ...)

WinCPP commented:
@karelz got pushed down the todo list, but I've cleared the decks. I'll do it in a day or two... everything is almost ready...

karelz commented:
Cool, thanks for update and rough ETA!

WinCPP commented on src/System.Drawing.Primitives/tests/SizeFTests.netcoreapp.cs:
```diff
@@ -23,19 +39,22 @@ public partial class SizeFTests
-            SizeF sz1 = new SizeF(value1, value1);
-            SizeF sz2 = new SizeF(value2, value2);
-            SizeF mulExpected;
+            MultSizeFFloatTester(value1, 1.0f, value2);
```
All the previous test cases for operator * or operator / treated width and height in symmetric way. Hence changed the test cases to test effect of operator on width and height effects separately.

While reverting the code for `operator /` for type `Size`, I mistakenly ended up writing `left._width / right` for both `width` and `height` params for creating new instance of `Size` returned from the `operator /`. That prompted me to make the above changes so that width and height are tested separately.

WinCPP commented:
@mellinoe Please review the changes. I have done some changes to the test cases to separately test the effects of operators on width and height components of type Size / SizeF.

WinCPP commented:
@dotnet-bot test Linux x64 Tests - Debug - Ubuntu.1404.Amd64.Open please

WinCPP commented:
@mellinoe @karelz I'm not able to figure out why the Linux test is failing. It mentions following which appears to be a bit unrelated... Kindly advise.

```
Unhandled Exception of Type Xunit.Sdk.EqualException
Message :
Assert.Equal() Failure\nExpected: 42\nActual:   134
Stack Trace :
   at System.Diagnostics.RemoteExecutorTestBase.RemoteInvokeHandle.Dispose() in /mnt/j/workspace/dotnet_corefx/master/portable-linux-Config_Debug+OuterLoop_false_prtest/src/Common/tests/System/Diagnostics/RemoteExecutorTestBase.cs:line 219
   at System.Data.SqlClient.Tests.DiagnosticTest.ExecuteScalarAsyncErrorTest() in /mnt/j/workspace/dotnet_corefx/master/portable-linux-Config_Debug+OuterLoop_false_prtest/src/System.Data.SqlClient/tests/FunctionalTests/DiagnosticTest.cs:line 280
```

mellinoe commented:
@dotnet-bot  Test Linux x64 Tests - Debug - Ubuntu.1404.Amd64.Open

@WinCPP That seems like [an unrelated error](https://github.com/dotnet/corefx/issues/17925). You can ignore it.

mellinoe commented:
@dotnet-bot Test Linux x64 Tests - Debug - Ubuntu.1404.Amd64.Open

mellinoe commented on src/System.Drawing.Primitives/tests/SizeFTests.netcoreapp.cs:
```diff
@@ -0,0 +1,87 @@
+            SizeF sz = new SizeF(width, height);
+            SizeF mulExpected;
+
+            unchecked
```
I don't think `unchecked` is needed here.

mellinoe commented on src/System.Drawing.Primitives/tests/SizeFTests.netcoreapp.cs:
```diff
@@ -0,0 +1,87 @@
+        [InlineData(1.0f, float.MinValue)]
+        [InlineData(-1.0f, float.MinValue)]
+        [InlineData(-1.0f, float.MinValue)]
+        public void DivideTestSizeFloat(float value1, float value2)
```
I think this should be called `DivideTestSizeFFloat` (two F's).

mmitche commented:
@mellinoe You can't restart just a single test leg yet.

@dotnet-bot test portable linux debug pipeline

mellinoe commented on src/System.Drawing.Primitives/tests/SizeFTests.netcoreapp.cs:
```diff
@@ -0,0 +1,84 @@
+        [InlineData(float.MinValue, float.MaxValue)]
+        public void MultiplicationTest(float value1, float value2)
+        {
+            MultSizeFFloatTester(value1, 1.0f, value2);
```
I find the structure of these tests a little unusual. For example, why are we only testing `SizeF` values which have either Width=1.0f or Height=1.0f?

I think I would prefer a "simpler" model, where the test case itself gets passed 5 values:
`public void MultiplyTest(float width, float height, float multiplier)`, and then does very simple assertions based on that. There isn't much to test here, IMO. We can probably have the same number of test cases and not lose much value. As long as we cover 1-2 "regular" cases, and then boundary behavior, we should be okay.

WinCPP replied:
Intention to have 0 and 1 is because they give different results, prominently, zero and identity. I don't see a reason to remove them either!

WinCPP replied:
@mellinoe @karelz can we please have all the review comments consolidated together rather than having to and fro? I will work on it in one go and push the changes.

WinCPP replied:
As for the code at line 39 and 40 - this is to test each of the components of Size / SizeF i..e width and height independently. I already mentioned that in one of the review comments as to what prompted me to do so. The arithmetic test cases handle width / height symmetrically. A typo of _width in place of _height or vice versa in the operator implementation would go undetected if the test cases use symmetric input to both components. Hence the idea to fix one of the two variables and test other.

WinCPP replied:
Please refer to the comment [here](https://github.com/dotnet/corefx/pull/20052#issuecomment-305262714) and [here](https://github.com/dotnet/corefx/pull/20052#discussion_r119422133)

mellinoe replied:
I think it is fine to have test cases which cover (width, 1f) or (1f, height); not arguing against that. But currently, all of the test cases are of that form. I think that we should at least have a test case which covers (3.0f, 4.0f) * 5.0f, for example.

My suggestion goes further than that, though. If the test case was specified like I mentioned above, I think the readability would also improve greatly. For example, it would be more obvious that `MutliplyTest(3f, 4f, 5f)` would be testing `new SizeF(3f, 4f) * 5f`. Right now I think it is difficult to identify which test parameters end up in where in the assertions, because they are used in a few different combinations.

Does that make sense? Part of this is the subjective "readability" of the test case, so other folks might disagree with my assessment.

WinCPP replied:
> a test case which covers (3.0f, 4.0f) * 5.0f, for example.
There is already a variation of that sort this line 26 as mentioned below.
```C#
       [InlineData(1000.234f, 2400.933f)]
```
Just that this works by instantiating two size objects (1000.234f, 1.0f) and (1.0f, 1000.234f) both being multiplied by 2400.933f. Having both varying at the same time is just another variation, but really not necessary when the individual components of Size/F object have been sufficiently covered.

As for 'readability', I think I would like to differ on that as the test cases are quite readable. There are far more complex test cases in this very repository wherein the test cases are based on intricate framework. One such being at [this link](https://github.com/dotnet/corefx/pull/18109/files) on which I worked!

So are we clear for accepting the changes or we want to seek opinion from wider audience?

Thanks Mandar

mellinoe replied:
The `Dictionary` type is far more complicated than this simple structure, so it only makes sense that it will have far more complicated test cases. I think there's just a few small things that could be cleaned up in the current version:

* `MultiplicationTest` has two vaguely-named parameters (value1 and value2), which are used in a few combinations of ways through a helper class `MultSizeFFloatTester`. If I add `InlineData` to the test case, I feel that it is difficult to immediately assess what the outcome of that will be without looking through all 4 call-sites. Instead, it would be clearer if `MultSizeFFloatTester` was inlined into `MutliplicationTest`, and all variants passed into it through `InlineData`. That involves less code paths to follow, less parameter combinations to grok, and will allow clearer parameter names to be used.
* Again, there are no test cases in which one of the fields is not 1.0f. It is true that, through our prior intuition about how the library code is written, this cannot matter. But from a test perspective it is odd, and hints that we are trying to verify some property that we are not. If you want to ensure that you are not returning `Width * multiplier` for both fields (you mentioned that), then you can just have a test case which passes different values for `width` and `height`.

I hope this doesn't sound needlessly pedantic. As the maintainer of the code, I need to make sure that this code is instantly understandable when I come back to it in several months (or longer), and if I or others need to add test cases to it, it should be as easy as possible for a type like this.

WinCPP replied:
Sounds like a solution... and I wouldn't agree more about being concerned as a maintainer of the code.

> MultiplicationTest has two vaguely-named ......... will allow clearer parameter names to be used
I was trying to avoid bloating the main function with various combinations. But please read next point...

>Again, there are no test .......... which passes different values for width and height.
This sounds like a solution. I was more concerned about the said typo that I almost did.

So I will revert back to the original test cases and add special test for different parameters, sort of (3.0f, 4.0f, 5.0f).

Between, I tried resolving a merge conflict and it seems it has messed something up. I am in process of fixing it up. Thanks!

WinCPP commented:
@mellinoe This commit was just to resolve the merge conflict errors and broken build. Please do not start the reviews yet.

WinCPP commented:
Hi @mellinoe I have done following changes,
1. Bring back the original test cases and added few more data points in each test case.
1. Rename value1 to dimension and value2 to multiplier / divisor. In such test cases the value of dimension is fed into width as well as height - symmetric inputs to Size/F components.
1. Added one three parameter - width, height, multiplier/divisor - test cases to each type of test scenario to check against repetition of width / height inside operator implementation - asymmetric inputs to Size/F components.

Kindly review and let me know if any changes. Its quite late here, will fix up tomorrow if any inputs. Thanks!

danmosemsft commented:
@mellinoe could you take anohter look to keep this moving along?

mellinoe commented:
@WinCPP Thanks for responding to all the feedback, and thanks for the contribution! Merging now.
